import grpc
from concurrent import futures
import time
import cv2
import numpy as np
import torch
import sys
sys.path.append("generated")
from generated import streaming_pb2, streaming_pb2_grpc

class VideoStreamingServicer(streaming_pb2_grpc.VideoStreamingServiceServicer):
    def __init__(self):
        self.vehicle_channel = grpc.insecure_channel("localhost:5001")
        self.vehicle_stub = streaming_pb2_grpc.VehicleDetectionServiceStub(self.vehicle_channel)
        self.plate_channel = grpc.insecure_channel("localhost:5002")
        self.plate_stub = streaming_pb2_grpc.PlateDetectionServiceStub(self.plate_channel)
        self.frame_count = 0  # Bi·∫øn ƒë·∫øm s·ªë l∆∞·ª£ng frame ƒë√£ nh·∫≠n
        self.last_frame_time = time.time()  # Th·ªùi gian nh·∫≠n frame cu·ªëi c√πng
    
    def StreamVideo(self, request_iterator, context):
        for frame in request_iterator:
            # TƒÉng b·ªô ƒë·∫øm frame
            self.frame_count += 1
            # Ki·ªÉm tra th·ªùi gian t·ª´ frame cu·ªëi c√πng
            current_time = time.time()
            if current_time - self.last_frame_time > 10:  # N·∫øu qu√° 10 gi√¢y
                print("‚ö†Ô∏è Kh√¥ng nh·∫≠n ƒë∆∞·ª£c frame trong 10 gi√¢y, reset frame_count v·ªÅ 0.")
                self.frame_count = 1  # Reset frame_count
            
            # C·∫≠p nh·∫≠t th·ªùi gian nh·∫≠n frame cu·ªëi c√πng
            self.last_frame_time = current_time

            

            print(f"Server: Received Frame ID: {frame.frame_id}")  # Ghi log frame_id
            vehicle_responses = self.vehicle_stub.DetectVehicles(iter([frame]))
            plate_responses = self.plate_stub.DetectPlates(iter([frame]))
            
            for vehicle_response in vehicle_responses:
                num_vehicles = len(vehicle_response.vehicles)
                print(f"Frame ID: {vehicle_response.frame_id}")
                print(f"  - S·ªë l∆∞·ª£ng ph∆∞∆°ng ti·ªán ph√°t hi·ªán: {num_vehicles}")
                
            for plate_response in plate_responses:
                num_plates = len(plate_response.plates)
                print(f"Frame ID: {plate_response.frame_id}")
                print(f"  - S·ªë l∆∞·ª£ng ph∆∞∆°ng ti·ªán ph√°t hi·ªán: {num_plates}")
            
            # Hi·ªÉn th·ªã t·ªïng s·ªë frame ƒë√£ nh·∫≠n
            print(f"T·ªïng s·ªë frame ƒë√£ nh·∫≠n: {self.frame_count}")
            
            yield frame

    # def StreamVideo(self, request_iterator, context):
    #     for frame in request_iterator:
    #         # Ki·ªÉm tra th·ªùi gian t·ª´ frame cu·ªëi c√πng
    #         current_time = time.time()
    #         if current_time - self.last_frame_time > 10:  # N·∫øu qu√° 10 gi√¢y
    #             print("‚ö†Ô∏è Kh√¥ng nh·∫≠n ƒë∆∞·ª£c frame trong 10 gi√¢y, reset frame_count v·ªÅ 0.")
    #             self.frame_count = 0  # Reset frame_count
    #         
    #         # C·∫≠p nh·∫≠t th·ªùi gian nh·∫≠n frame cu·ªëi c√πng
    #         self.last_frame_time = current_time
    #         
    #         # TƒÉng b·ªô ƒë·∫øm frame
    #         self.frame_count += 1
    #         
    #         print(f"Server: Received Frame ID: {frame.frame_id}")  # Ghi log frame_id
    #         vehicle_responses = self.vehicle_stub.DetectVehicles(iter([frame]))
    #         plate_responses = self.plate_stub.DetectPlates(iter([frame]))
    #         
    #         for vehicle_response, plate_response in zip(vehicle_responses, plate_responses):
    #             num_vehicles = len(vehicle_response.vehicles)
    #             num_plates = len(plate_response.plates)
    #             print(f"Frame ID: {vehicle_response.frame_id}")
    #             print(f"  - S·ªë l∆∞·ª£ng ph∆∞∆°ng ti·ªán ph√°t hi·ªán: {num_vehicles}")
    #             print(f"  - S·ªë l∆∞·ª£ng bi·ªÉn s·ªë ph√°t hi·ªán: {num_plates}")
    #         
    #         # Hi·ªÉn th·ªã t·ªïng s·ªë frame ƒë√£ nh·∫≠n
    #         print(f"T·ªïng s·ªë frame ƒë√£ nh·∫≠n: {self.frame_count}")
    #         
    #         yield frame
            
def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=20))
    streaming_pb2_grpc.add_VideoStreamingServiceServicer_to_server(VideoStreamingServicer(), server)
    server.add_insecure_port("[::]:5000")
    server.start()
    print("üöÄ gRPC Video Streaming Server ƒëang ch·∫°y tr√™n c·ªïng 5000...")
    server.wait_for_termination()

if __name__ == "__main__":
    serve()
